src/sync/sync_session.cpp

```diff
@@ -404,6 +404,48 @@ void SyncSession::create_sync_session()
         }
     };
     m_session->set_sync_transact_callback(std::move(wrapped_callback));
+
+    // Set up the wrapped progress handler callback
+    auto wrapped_progress_handler = [this, weak_self](uint_fast64_t downloaded,
+                                                      uint_fast64_t downloadable,
+                                                      uint_fast64_t uploaded,
+                                                      uint_fast64_t uploadable) {
+        auto self = weak_self.lock();
+        if (!self) {
+            // An update was delivered after the session it relates to was destroyed. There's nothing useful
+            // we can do with it.
+            return;
+        }
+        std::lock_guard<std::mutex> lock(m_progress_notifier_mutex);
+        m_current_uploadable = uploadable;
+        m_current_downloadable = downloadable;
+        m_current_downloaded = downloaded;
+        m_current_uploaded = uploaded;
+        // Invoke all registered notifiers.
+        std::vector<uint64_t> notifiers_to_remove;
+        for (auto& pair : m_notifiers) {
+            auto& package = pair.second;
+            if (package.is_streaming) {
+                package.notifier(downloaded, downloadable, uploaded, uploadable);
```

Comment:
I talked with Austin about the fact that we're calling the notification callbacks with a lock held, and the potential for deadlocks if the callback wants to register or unregister progress notifications. We can avoid this by moving the calls to the callbacks outside the scope of the lock.